File: visa-scheduler-java/src/main/java/com/visascheduler/AisClient.java
package com.visascheduler;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.hc.client5.http.classic.methods.HttpGet;
import org.apache.hc.client5.http.classic.methods.HttpPost;
import org.apache.hc.client5.http.cookie.StandardCookieSpec;
import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.client5.http.impl.cookie.RFC6265CookieSpecFactory;
import org.apache.hc.core5.http.NameValuePair;
import org.apache.hc.core5.http.message.BasicNameValuePair;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * HTTP client that talks to the AIS (Appointment Information System) portal.
 * Maintains cookies across requests so the session stays authenticated.
 */
public class AisClient implements AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(AisClient.class);
    private static final Pattern SCHEDULE_ID_PATTERN = Pattern.compile("/schedule/(\\d+)");
    private static final ObjectMapper JSON = new ObjectMapper();

    private final CloseableHttpClient http;
    private String scheduleId;

    public AisClient() {
        this.http = HttpClients.custom()
                .setDefaultRequestConfig(
                        org.apache.hc.client5.http.config.RequestConfig.custom()
                                .setResponseTimeout(
                                        org.apache.hc.core5.util.Timeout.ofSeconds(30))
                                .setConnectionRequestTimeout(
                                        org.apache.hc.core5.util.Timeout.ofSeconds(30))
                                .setCookieSpec(StandardCookieSpec.RELAXED)
                                .build())
                .build();
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    /**
     * Log in to AIS and populate {@code scheduleId}.
     *
     * @throws IOException          on network error
     * @throws RuntimeException     on login failure
     */
    public void login() throws IOException {
        log.info("Logging in as {} …", Config.AIS_EMAIL);

        String csrf = fetchCsrf(Config.SIGN_IN_URL);

        List<NameValuePair> form = List.of(
                new BasicNameValuePair("utf8",                   "✓"),
                new BasicNameValuePair("authenticity_token",     csrf),
                new BasicNameValuePair("user[email]",            Config.AIS_EMAIL),
                new BasicNameValuePair("user[password]",         Config.AIS_PASSWORD),
                new BasicNameValuePair("policy_confirmed",       "1"),
                new BasicNameValuePair("commit",                 "Sign In")
        );

        HttpPost post = buildPost(Config.SIGN_IN_URL, form);
        post.setHeader("Referer", Config.SIGN_IN_URL);

        try (CloseableHttpResponse resp = http.execute(post)) {
            String body = new String(resp.getEntity().getContent().readAllBytes(),
                    StandardCharsets.UTF_8);
            String finalUrl = extractFinalUrl(resp, Config.SIGN_IN_URL);

            if (finalUrl.contains("sign_in")) {
                throw new RuntimeException(
                        "Login failed — check AIS_EMAIL/AIS_PASSWORD. Final URL: " + finalUrl);
            }

            scheduleId = extractScheduleId(finalUrl, body);
            log.info("Logged in successfully. Schedule ID: {}", scheduleId);
        }
    }

    private String extractFinalUrl(CloseableHttpResponse resp, String fallback) {
        var location = resp.getFirstHeader("Location");
        return (location != null) ? location.getValue() : fallback;
    }

    private String extractScheduleId(String url, String html) {
        Matcher m = SCHEDULE_ID_PATTERN.matcher(url);
        if (m.find()) return m.group(1);

        Document doc = Jsoup.parse(html);
        for (Element a : doc.select("a[href]")) {
            Matcher am = SCHEDULE_ID_PATTERN.matcher(a.attr("href"));
            if (am.find()) return am.group(1);
        }
        throw new RuntimeException("Could not determine schedule ID after login.");
    }

    // ── Slot Queries ──────────────────────────────────────────────────────────

    /** Returns list of available appointment dates (YYYY-MM-DD). */
    public List<String> getAvailableDates() throws IOException {
        String url = Config.availableDatesUrl(scheduleId);
        String json = getJson(url);

        List<String> dates = new ArrayList<>();
        JsonNode root = JSON.readTree(json);
        if (root.isArray()) {
            for (JsonNode node : root) {
                if (node.has("date")) {
                    dates.add(node.get("date").asText());
                }
            }
        }
        return dates;
    }

    /** Returns available time slots (e.g. "08:30") for a given date. */
    public List<String> getAvailableTimes(String date) throws IOException {
        String url = Config.availableTimesUrl(scheduleId, date);
        String json = getJson(url);

        List<String> times = new ArrayList<>();
        JsonNode root = JSON.readTree(json);
        JsonNode available = root.path("available_times");
        if (available.isArray()) {
            for (JsonNode t : available) {
                times.add(t.asText());
            }
        }
        return times;
    }

    // ── Booking ───────────────────────────────────────────────────────────────

    /**
     * Submit the appointment booking form.
     *
     * @return true if the booking was confirmed
     */
    public boolean bookAppointment(String date, String time) throws IOException {
        String apptUrl = Config.appointmentsUrl(scheduleId);

        String csrf;
        try {
            csrf = fetchCsrf(apptUrl);
        } catch (Exception e) {
            log.error("Could not fetch CSRF before booking: {}", e.getMessage());
            return false;
        }

        List<NameValuePair> form = List.of(
                new BasicNameValuePair("utf8",                                          "✓"),
                new BasicNameValuePair("authenticity_token",                            csrf),
                new BasicNameValuePair("confirmed_limit_message",                       "1"),
                new BasicNameValuePair("use_consulate_appointment_capacity",            "true"),
                new BasicNameValuePair("appointments[consulate_appointment][facility_id]",
                        String.valueOf(Config.FACILITY_ID)),
                new BasicNameValuePair("appointments[consulate_appointment][date]",     date),
                new BasicNameValuePair("appointments[consulate_appointment][time]",     time)
        );

        HttpPost post = buildPost(apptUrl, form);
        post.setHeader("Referer", apptUrl);

        try (CloseableHttpResponse resp = http.execute(post)) {
            String body = new String(resp.getEntity().getContent().readAllBytes(),
                    StandardCharsets.UTF_8);
            String finalUrl = extractFinalUrl(resp, apptUrl);

            int status = resp.getCode();
            if (status == 200 && body.contains("Appointment Confirmed")) {
                log.info("Appointment confirmed on page: {} at {}", date, time);
                return true;
            }
            if (!finalUrl.contains("new") && finalUrl.contains("appointment")) {
                log.info("Booking likely succeeded — redirected to: {}", finalUrl);
                return true;
            }

            log.warn("Booking response HTTP {} — final URL: {}", status, finalUrl);
            return false;
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String fetchCsrf(String url) throws IOException {
        HttpGet get = buildGet(url);
        try (CloseableHttpResponse resp = http.execute(get)) {
            String html = new String(resp.getEntity().getContent().readAllBytes(),
                    StandardCharsets.UTF_8);
            Document doc = Jsoup.parse(html);
            Element meta = doc.selectFirst("meta[name=csrf-token]");
            if (meta == null) {
                throw new RuntimeException("CSRF token not found on: " + url);
            }
            return meta.attr("content");
        }
    }

    private String getJson(String url) throws IOException {
        HttpGet get = buildGet(url);
        get.setHeader("Accept",           "application/json, text/javascript, */*; q=0.01");
        get.setHeader("X-Requested-With", "XMLHttpRequest");
        try (CloseableHttpResponse resp = http.execute(get)) {
            return new String(resp.getEntity().getContent().readAllBytes(),
                    StandardCharsets.UTF_8);
        }
    }

    private HttpGet buildGet(String url) {
        HttpGet get = new HttpGet(url);
        get.setHeader("User-Agent",       Config.USER_AGENT);
        get.setHeader("Accept-Language",  "en-US,en;q=0.9");
        get.setHeader("Referer",          Config.BASE_URL);
        return get;
    }

    private HttpPost buildPost(String url, List<NameValuePair> form) {
        HttpPost post = new HttpPost(url);
        post.setHeader("User-Agent",      Config.USER_AGENT);
        post.setHeader("Accept-Language", "en-US,en;q=0.9");
        post.setEntity(new UrlEncodedFormEntity(form, StandardCharsets.UTF_8));
        return post;
    }

    @Override
    public void close() throws IOException {
        http.close();
    }
}

File: visa-scheduler/ais_client.py
"""
AIS (Appointment Information System) HTTP client.

Uses a requests.Session to maintain cookies/CSRF tokens across calls,
exactly as the real browser does when navigating the AIS website.
"""

from __future__ import annotations

import logging
import re
import time
from datetime import date, datetime
from typing import Optional

import requests
from bs4 import BeautifulSoup

import config

logger = logging.getLogger(__name__)

_DATE_FMT = "%Y-%m-%d"
_TIME_FMT = "%I:%M %p"


class AISClient:
    """Thin wrapper around the AIS REST + HTML endpoints."""

    def __init__(self) -> None:
        self.session = requests.Session()
        self.session.headers.update(
            {
                "User-Agent": config.USER_AGENT,
                "Accept-Language": "en-US,en;q=0.9",
                "Referer": config.BASE_URL,
            }
        )
        self.schedule_id: Optional[str] = None
        self._csrf: Optional[str] = None

    # ──────────────────────────────────────────────
    # Auth
    # ──────────────────────────────────────────────

    def _get_csrf(self, url: str) -> str:
        """Fetch a page and extract the Rails CSRF token."""
        resp = self.session.get(url, timeout=30)
        resp.raise_for_status()
        soup = BeautifulSoup(resp.text, "html.parser")
        tag = soup.find("meta", {"name": "csrf-token"})
        if tag is None:
            raise RuntimeError("CSRF token not found on page.")
        return tag["content"]

    def login(self) -> None:
        """Authenticate with AIS and populate self.schedule_id."""
        logger.info("Logging in as %s …", config.AIS_EMAIL)
        csrf = self._get_csrf(config.SIGN_IN_URL)

        payload = {
            "utf8": "✓",
            "authenticity_token": csrf,
            "user[email]": config.AIS_EMAIL,
            "user[password]": config.AIS_PASSWORD,
            "policy_confirmed": "1",
            "commit": "Sign In",
        }
        resp = self.session.post(
            config.SIGN_IN_URL,
            data=payload,
            headers={"Referer": config.SIGN_IN_URL},
            timeout=30,
            allow_redirects=True,
        )
        resp.raise_for_status()

        if "sign_in" in resp.url:
            raise RuntimeError(
                "Login failed — check AIS_EMAIL / AIS_PASSWORD. "
                f"Landed on: {resp.url}"
            )

        self.schedule_id = self._extract_schedule_id(resp.url, resp.text)
        logger.info("Logged in. Schedule ID: %s", self.schedule_id)

    def _extract_schedule_id(self, url: str, html: str) -> str:
        """Pull the numeric schedule id from the redirect URL or page links."""
        # The URL usually looks like /en-ca/niv/schedule/12345678/continue
        match = re.search(r"/schedule/(\d+)", url)
        if match:
            return match.group(1)

        # Fall back to parsing page links
        soup = BeautifulSoup(html, "html.parser")
        for a in soup.find_all("a", href=True):
            m = re.search(r"/schedule/(\d+)", a["href"])
            if m:
                return m.group(1)

        raise RuntimeError("Could not determine schedule ID after login.")

    # ──────────────────────────────────────────────
    # Slot Queries
    # ──────────────────────────────────────────────

    def get_available_dates(self) -> list[str]:
        """Return a list of available appointment dates (YYYY-MM-DD strings)."""
        url = config.AVAILABLE_DATES_URL.format(schedule_id=self.schedule_id)
        resp = self.session.get(
            url,
            headers={
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "X-Requested-With": "XMLHttpRequest",
                "Referer": config.APPOINTMENTS_URL.format(
                    schedule_id=self.schedule_id
                ),
            },
            timeout=30,
        )
        resp.raise_for_status()
        data = resp.json()
        # Each element is {"date": "2026-07-15", "business_day": true}
        dates = [entry["date"] for entry in data if isinstance(entry, dict)]
        return dates

    def get_available_times(self, appt_date: str) -> list[str]:
        """Return available time slots for a given date."""
        url = config.AVAILABLE_TIMES_URL.format(
            schedule_id=self.schedule_id, date=appt_date
        )
        resp = self.session.get(
            url,
            headers={
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "X-Requested-With": "XMLHttpRequest",
                "Referer": config.APPOINTMENTS_URL.format(
                    schedule_id=self.schedule_id
                ),
            },
            timeout=30,
        )
        resp.raise_for_status()
        data = resp.json()
        # Response: {"available_times": ["08:30", ...], "business_times": [...]}
        return data.get("available_times", [])

    # ──────────────────────────────────────────────
    # Booking
    # ──────────────────────────────────────────────

    def book_appointment(self, appt_date: str, appt_time: str) -> bool:
        """
        Submit the appointment booking form.
        Returns True on success, False on failure.
        """
        appt_url = config.APPOINTMENTS_URL.format(schedule_id=self.schedule_id)

        # Refresh CSRF token from the appointment page
        try:
            csrf = self._get_csrf(appt_url)
        except Exception as exc:
            logger.error("Could not fetch CSRF before booking: %s", exc)
            return False

        payload = {
            "utf8": "✓",
            "authenticity_token": csrf,
            "confirmed_limit_message": "1",
            "use_consulate_appointment_capacity": "true",
            "appointments[consulate_appointment][facility_id]": str(config.FACILITY_ID),
            "appointments[consulate_appointment][date]": appt_date,
            "appointments[consulate_appointment][time]": appt_time,
        }

        resp = self.session.post(
            appt_url,
            data=payload,
            headers={"Referer": appt_url},
            timeout=30,
            allow_redirects=True,
        )

        if resp.status_code == 200 and "Appointment Confirmed" in resp.text:
            logger.info("Appointment booked: %s at %s", appt_date, appt_time)
            return True

        # Check for a success redirect pattern
        if "appointment" in resp.url and "new" not in resp.url:
            logger.info(
                "Booking likely succeeded (redirected to %s).", resp.url
            )
            return True

        logger.warning(
            "Booking response status %s. URL: %s", resp.status_code, resp.url
        )
        return False
